skops/io/_visualize.py

```diff
@@ -0,0 +1,262 @@
+from __future__ import annotations
+
+import json
+from dataclasses import dataclass
+from functools import singledispatch
+from pathlib import Path
+from typing import Callable, Iterator, Literal, Sequence
+from zipfile import ZipFile
+
+from ._audit import Node, get_tree
+from ._general import FunctionNode, JsonNode
+from ._numpy import NdArrayNode
+from ._scipy import SparseMatrixNode
+from ._utils import LoadContext
+
+
+@dataclass
+class PrintConfig:
+    # fmt: off
+    tag_safe: str =   ""  # noqa: E222
+    tag_unsafe: str = " [UNSAFE]"
+
+    line_start: str = "├─"
+    line: str =       "──"  # noqa: E222
+
+    use_colors: bool = True
+    color_safe: str =         '\033[32m'  # green  # noqa: E222
+    color_unsafe: str =       '\033[31m'  # red  # noqa: E222
+    color_child_unsafe: str = '\033[33m'  # yellow
+    color_end: str =          '\033[0m'   # noqa: E222
+    # fmt: on
+
+
+print_config = PrintConfig()
+
+
+@dataclass
+class FormattedNode:
+    level: int
+    key: str  # the key to the node
+    val: str  # the value of the node
+    visible: bool  # whether it should be shown
+
+
+def pretty_print_tree(
+    formatted_nodes: Iterator[FormattedNode], config: PrintConfig
+) -> None:
+    # TODO: the "tree" lines could be made prettier since all nodes are known
+    # here
+    for formatted_node in formatted_nodes:
+        if not formatted_node.visible:
+            continue
+
+        line = print_config.line_start
+        line += (formatted_node.level - 1) * print_config.line
+        line += f"{formatted_node.key}: {formatted_node.val}"
+        print(line)
+
+
+def _check_visibility(
+    node: Node,
+    node_is_safe: bool,
+    node_and_children_are_safe: bool,
+    show: Literal["all", "untrusted", "trusted"],
+) -> bool:
+    if show == "all":
+        should_print = True
+    elif show == "untrusted":
+        should_print = not node_and_children_are_safe
+    else:  # only trusted
+        should_print = node_is_safe
+    return should_print
+
+
+def _check_node_and_children_safe(node: Node, trusted: bool | Sequence[str]) -> bool:
```

Comment:
I think it'd be okay to move this to nodes and cache them.

Reply:
Oh yeah, there is already `node.is_safe()` >__<